doc/ztp/ztp_hld.md

```diff
@@ -0,0 +1,166 @@
+- Updategraph handles all configuration during ZTP service
+- ZTP should receive post config validation script, that collects switch info and sends it back to a remote location for processing
+- ZTP status should be logged through syslog
+- Interruption of ZTP service should set ZTP status as incomplete/interrupted and ZTP should be disabled.
+
+# DHCP Options
+The following are the private DHCP options used during the implementation of updategraph/ZTP
+
+| DHCP Option | Information         | Explanation                                       |
+|:-----------:|:-------------------|:-------------------------------------------------|
+|    224      | snmp_community     | usage implemented through updategraph |
+|    225      | minigraph_url     | usage implemented through updategraph |
+|    226      | acl_url           | usage implemented through updategraph |
+|    227      | ztp_image_url     | URL for the SONiC image for the switch to dowload |
+|    228      | validation_url    | URL for the validation script -- a script that runs post config and  collects device info and sends it for processing to a remote location designated within the validation script |
+
+# Flow
+![](https://github.com/simone-dell/SONiC/blob/ztp/images/ztp_hld/ztp_flow.jpg)
+
+For ZTP using DHCP, provisioning initially takes place over the management network and is initiated through a DHCP hook. A DHCP option is used to specify a configuration script. This script is then requested from the Web server and executed locally on the switch.
+1.	Simplify installation of switch, the steps involved will be
+    - Rack and Stack 
+    - Connect 
+    - Power-on 
+2.	Onie Boots into SONiC (with ZTP enabled) from flash
+    - Power-on ZTP service will take over to load the initial configuration.
+    - ZTP service flag $enabled is true
+    - ZTP service flag $post_install is false
+    - ZTP service flag $post_config is false
+3.	The switch now enters ZTP mode and does the following,
+    - Obtains an IP address from the DHCP server
+    - Obtains the URL for the SONIC image, minigraph, and post config validation script 
+4.	At this step, the switch will have information to reach the HTTP / TFTP server through URLs given by DHCP server.
+5.	ZTP service will enable and start updategraph service
+    - Within updategraph, ztp_enabled is true, and post_install is false
+    - This triggers updategraph to configure the device with an default config
+    - Upon exiting updategraph, SNMP will start
+6.	ZTP service will now download the software image file.
+7.	sonic_installer will validate image to be a SONIC image compatible with the current platform.
+8.	sonic_installer  will check remaining hard disk for space before doing image install.
+9.	ZTP service will install image using sonic-installer
+    - ZTP service flag $post_install is true
+10.	ZTP service will enable updategraph
+11.	ZTP service will call for reboot to apply new SONIC image, with option ZTP-enabled –post_install
```

Comment:
Can we avoid this re-boot? It doesn't seem necessary, especially in cases where the image install happens in ONIE.